docs/src/ref/distributions.md

```diff
@@ -19,3 +17,207 @@ exponential
 laplace
 ```
 
+## Creating New Distributions via the Distributions DSL
+
+In addition to using the above built-in distributions, and to defining subtypes
+of `Gen.Distribution` in plain Julia, you may combine existing distributions to
+create new ones using the `@dist` DSL.  The syntax for this DSL is
+
+```julia
+@dist name(arg1, arg2, ..., argN) = body
+```
+or
+```julia
+@dist function name(arg1, arg2, ..., argN)
+    body
+end
+```
+
+Here `body` is ordinary Julia code, with the constraint that `body` must
+contain exactly one random choice.  The value of the `@dist` expression is then
+a `Gen.distribution` object called `name`, parameterized by `arg1, ..., argN`,
+representing the distribution over _return values_ of `body`.
+
+This DSL is designed to address the issue that sometimes, values stored in the
+trace do not correspond to the most natural physical elements of the model
+state space, making inference programming and querying more taxing than
+necessary. For example, suppose we have a model of classes at a school, where
+the number of students is random, with mean 10, but always at least 3. Rather
+than writing the model as
+
+```julia
+@gen function class_model()
+   n_students = @trace(poisson(7), :n_students_minus_3) + 3
+   ...
+end
+```
+
+and thinking about the random variable `:n_students_minus_3`, you can use the
+`@dist` DSL to instead write
+
+```julia
+@dist student_distr(mean, min) = poisson(mean-min) + min
+
+@gen function class_model()
+   n_students = @trace(student_distr(10, 3), :n_students)
+   ...
+end
+```
+
+and think about the more natural random variable `:n_students`.  This leads to
+more natural inference programs, which can constrain and propose directly to
+the `:n_students` trace address.
+
+### Permitted means of combination
+
+Of course, it is not possible for `@dist` to work on any arbitrary `body`.  We
+now describe which constructs are permitted inside the `body` of a `@dist`
+expression.
+
+We can think of the `body` of an `@dist` function as containing ordinary Julia
+code, except that in addition to being described by their ordinary Julia types,
+each expression also belongs to one of three "type spaces." These are:
+
+1. `CONST`: Constants, whose value is known at the time this `@dist` expression
+   is evaluated.
+2. `ARG`: Arguments and (deterministic, differentiable) functions of arguments.
+   All expressions representing non-random values that depend on distribution
+   arguments are `ARG` expressions.
+3. `RND`: Random variables. All expressions whose runtime values may differ
+   across multiple calls to this distribution (with the same arguments) are
+   `RND` expressions.
+
+**Importantly, Julia control flow constructs generally expect `CONST` values:
+the condition of an `if` or the range of a `for` loop cannot be `ARG` or
+`RND`.**
+_(Developer note: We could change this: it would not be too difficult, e.g., to
```

Comment:
I'd be in favor of removing the "developer note" here. If we want to make this a priority, we can open an issue?

Reply:
Good call, an issue is the right place for this.  It's also probably the right place for the "long-term future note" below.  Filed https://github.com/probcomp/Gen/issues/199.